private/syntax-replacement.rkt

```diff
@@ -295,6 +301,35 @@
        (check-equal? (syntax-replacement-render replacement) expected)])))
 
 
+(define (syntax-replacement-preserves-free-identifiers? replacement)
+  (match replacement
+    [(syntax-replacement #:original-syntax orig
+                         #:new-syntax new
+                         #:introduction-scope intro)
+     ((stx-free-ids=?/ignore
+       (ignore-id? intro (list #'SPACE #'NEWLINE #'ORIGINAL-GAP #'ORIGINAL-SPLICE)))
+      new
+      (datum->syntax orig (syntax->datum new)))]))
+
+;; ignore-id? : (Syntax -> Syntax) (Listof Id) -> Id -> Boolean
+(define ((ignore-id? scope ignore) id)
+  (or (member id ignore free-identifier=?)
+      (bound-identifier=? id (scope id 'remove))))
+
+;; stx-e : Stx -> Any
+(define (stx-e stx) (if (syntax? stx) (syntax-e stx) stx))
+
+;; stx-free-ids=?/only : (Id -> Boolean) -> Stx Stx -> Boolean
+(define (stx-free-ids=?/ignore ignore?)
+  (define (stx-free-ids=? a b)
+    (cond
+      [(and (identifier? a) (identifier? b))
+       (or (free-identifier=? a b) (and (ignore? a) (ignore? b)))]
+      [else
+       (equal?/recur (stx-e a) (stx-e b) stx-free-ids=?)]))
+  stx-free-ids=?)
```

Comment:
I prefer to avoid relying on `equal?/recur`, it's a little too brittle in my opinion. There's an `in-syntax-identifiers` function I wrote in `resyntax/default-recommendations/private/syntax-identifier-sets` that you can use to implement this logic instead:

```suggestion
;; stx-free-ids=?/ignore : (Id -> Boolean) -> Stx Stx -> Boolean
(define ((stx-free-ids=?/ignore ignore?) stx1 stx2)
  (for/and ([id1 (in-syntax-identifiers stx1)]
            [id2 (in-syntax-identifiers stx2)])
    (or (and (ignore? id1) (ignore? id2)) (free-identifier=? id1 id2))))
```

This also removes the need for recursion. In fact, it gets simple enough that it might be worth just inlining it if it's only used in `syntax-replacement-preserves-free-identifiers?`. Inlining it and the `ignore-id?` predicate would make the purpose of this check easier to understand.

Reply:
I'd rather not use two `in-syntax-identifiers` sequences in parallel on two syntax objects that just so "happen" to have the same tree structure. But this does give me a better idea: I can avoid constructing the second syntax object and just traverse the identifiers in the first!